### PR TITLE
Pods which exit and won't restart should not be in the Endpoints.NotReadyAddresses

### LINKED ISSUE #46776: Failed pods with restart policy=Never should not be in the Endpoints.NotReadyAddresses
If you create a job in which pods are crashing and a service spawning those pods, this may result in generating extremely huge Endpoints object (which will cause different performance issues).
[Based on situation from one of our customers cluster].

After 30m from creating the objects below in my 3 node-cluster I have an endpoint which has almost 1MB of size:
```
wojtekt@wojtekt-work:~/go/src/k8s.io/kubernetes$ kubectl get endpoints failed-service -o json > out.txt
wojtekt@wojtekt-work:~/go/src/k8s.io/kubernetes$ ls -la out.txt 
-rw-r----- 1 wojtekt eng 891834 Jun  1 14:31 out.txt
```

Example objects which are enough to cause this
1. service:
```
apiVersion: v1
kind: Service
metadata:
  name: failed-service
spec:
  selector:
    app: failed
  ports:
  - protocol: TCP
    port: 80
    targetPort: 12345
```
2. job:
```
apiVersion: batch/v1
kind: Job
metadata:
  name: failed-job2
spec:
  parallelism: 10
  completions: 10
  template:
    metadata:
      name: failed-job2
      labels:
        app: failed
    spec:
      restartPolicy: Never
      containers:
      - name: failed-pod
        image: busybox
        command: [ "/bin/sh", "-c", "sleep 5 && exit 1"]
```

@bowei @kubernetes/sig-network-bugs @kubernetes/sig-scalability-bugs

### PR DESCRIPTION
**Pull request by xiangpengzhao**
**What this PR does / why we need it**:
Such pods should not be in the Endpoints.NotReadyAddresses:
* Failed pods with restart policy=Never
* Succeeded pods with restart policy=Never
* Succeeded pods with restart policy=OnFailure

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #46776

**Special notes for your reviewer**:
/cc @wojtek-t 

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by cmluciano** [comment]
@k8s-bot ok to test

File: pkg/controller/endpoint/endpoints_controller.go [hunk 1]
```diff
@@ -415,6 +415,11 @@ func (e *EndpointController) syncService(key string) error {
 				})
 				readyEps++
 			} else {
+				if (pod.Spec.RestartPolicy == v1.RestartPolicyNever && pod.Status.Phase == v1.PodFailed) ||
```
Comment (wojtek-t) [inline pkg/controller/endpoint/endpoints_controller.go]: Can you extract those conditions to a function, like shouldPodBeInEndpoints (better name would be awesome)?

**Comment by xiangpengzhao** [comment]
@wojtek-t inline comment addressed, PTAL. I don't find a better func name, so I use the name you suggested:) 
Thanks!

File: pkg/controller/endpoint/endpoints_controller.go [hunk 2]
```diff
@@ -525,3 +525,12 @@ func addEndpointSubset(subsets []v1.EndpointSubset, pod *v1.Pod, epa v1.Endpoint
 	}
 	return subsets, readyEps, notReadyEps
 }
+
+func shouldPodBeInEndpoints(pod *v1.Pod) bool {
+	if (pod.Spec.RestartPolicy == v1.RestartPolicyNever && pod.Status.Phase == v1.PodFailed) ||
```
Comment (wojtek-t) [inline pkg/controller/endpoint/endpoints_controller.go]: Can we structure it in slightly more readable way, like:
```
switch pod.Spec.RestartPolicy {
case v1.RestartPolicyNever:
  return pod.Status.Phase != v1.PodFailed && pod.Status.Phase != v1.PodSucceeded
case v1.RestartPolicyOnFailure:
  return pod.Status.Phase != v1.PodSucceeded
default:
  return true
}
```

Comment (xiangpengzhao) [inline pkg/controller/endpoint/endpoints_controller.go]: Aha, thanks! Will do. I thought to use switch/case, but I was faint when I rebased the code:) I'd also like to write test case for `shouldPodBeInEndpoints`. Worthy?

Comment (wojtek-t) [inline pkg/controller/endpoint/endpoints_controller.go]: Having a unit test is always good.

**Comment by xiangpengzhao** [comment]
@wojtek-t PTAL. Thanks!

**Pull request merged**